File: IOS/KiddieApps/KiddieApps/DataLayer.swift
//
//  DataLayer.swift
//  KiddieApps
//
//  Created by Aloysius Ari Wicaksono on 6/20/17.
//  Copyright © 2017 Samanasoft. All rights reserved.
//

import Foundation


public class Announcement : BaseClass{
    var AnnouncementID:NSNumber?
    var Title:String?
    var StartDate:DateTime?
    var EndDate:DateTime?
    var GCAnnouncementType:String?
    var AnnouncementType:String?
    var Remarks:String?
    var LastUpdatedDate:DateTime?
    
    override func getPrimaryKey() -> [String]{
        return ["AnnouncementID"];
    }
}
public class AnnouncementDao{
    private var helper:DBHelper;
    private let p_AnnouncementID = "@p_AnnouncementID";
    
    public init() {
        helper = DBHelper();
    }
    public func get(AnnouncementID:Int) -> Announcement?{
        sharedInstance.database!.open()
        var query = helper.getRecord(tableName: "Announcement", lstPrimaryKey: Announcement().getPrimaryKey());
        query = query.replacingOccurrences(of: p_AnnouncementID, with: String(AnnouncementID));
        let row = DaoBase.getInstance().getDataRow(query: query);
        let temp = helper.dataListRowToObject(row: row, obj: Announcement());
        sharedInstance.database!.close()
        if(temp != nil) {
            let result = temp as! Announcement?;
            return result;
        }
        return nil
    }
    public func insert(record:Announcement) -> Bool{
        let query = helper.insert(tableName: "Announcement", record: record);
        return DaoBase.getInstance().executeNonQuery(query: query!);
    }
    public func update(record:Announcement) -> Bool{
        let query = helper.update(tableName: "Announcement", record: record);
        return DaoBase.getInstance().executeNonQuery(query: query!);
    }
    public func delete(AnnouncementID:Int) -> Bool{
        let record = get(AnnouncementID: AnnouncementID);
        if(record != nil){
            let query = helper.delete(tableName: "Announcement", record: record!);
            return DaoBase.getInstance().executeNonQuery(query: query!);
        }
        return false;
    }
}


public class Appointment : BaseClass{
    var AppointmentID:NSNumber?
    var MRN:NSNumber?
    var ServiceUnitName:String?
    var QueueNo:NSNumber?
    var StartDate:DateTime?
    var ReminderDate:DateTime?
    var EndDate:DateTime?
    var StartTime:String?
    var EndTime:String?
    var cfStartTime:String?
    var VisitTypeName:String?
    var ParamedicName:String?
    var SpecialtyName:String?
    var GCAppointmentStatus:String?
    var LastUpdatedDate:DateTime?
    
    override func getPrimaryKey() -> [String]{
        return ["AppointmentID"];
    }
}
public class AppointmentDao{
    private var helper:DBHelper;
    private let p_AppointmentID = "@p_AppointmentID";
    
    public init() {
        helper = DBHelper();
    }
    public func get(AppointmentID:Int) -> Appointment?{
        sharedInstance.database!.open()
        var query = helper.getRecord(tableName: "Appointment", lstPrimaryKey: Appointment().getPrimaryKey());
        query = query.replacingOccurrences(of: p_AppointmentID, with: String(AppointmentID));
        let row = DaoBase.getInstance().getDataRow(query: query);
        let temp = helper.dataListRowToObject(row: row, obj: Appointment());
        sharedInstance.database!.close()
        if(temp != nil) {
            let result = temp as! Appointment?;
            return result;
        }
        return nil
    }
    public func insert(record:Appointment) -> Bool{
        let query = helper.insert(tableName: "Appointment", record: record);
        return DaoBase.getInstance().executeNonQuery(query: query!);
    }
    public func update(record:Appointment) -> Bool{
        let query = helper.update(tableName: "Appointment", record: record);
        return DaoBase.getInstance().executeNonQuery(query: query!);
    }
    public func delete(AppointmentID:Int) -> Bool{
        let record = get(AppointmentID: AppointmentID);
        let query = helper.delete(tableName: "Appointment", record: record!);
        return DaoBase.getInstance().executeNonQuery(query: query!);
    }
}

public class LaboratoryResultDt : BaseClass{
    var LaboratoryResultDtID:NSNumber?
    var ID:NSNumber?
    var MRN:NSNumber?
    var ItemName1:String?
    var ItemGroupName1:String?
    var LabTestResultTypeID:NSNumber?
    var LabTestResultTypeName:String?
    var FractionID:NSNumber?
    var FractionName1:String?
    var MetricResultValue:NSNumber?
    var MinMetricNormalValue:NSNumber?
    var MaxMetricNormalValue:NSNumber?
    var MetricUnit:String?
    var NilaiRujukanMetric:String?
    var InternationalResultValue:NSNumber?
    var MinInternationalNormalValue:NSNumber?
    var MaxInternationalNormalValue:NSNumber?
    var InternationalUnit:String?
    var NilaiRujukanInternational:String?
    var LabTestResultTypeDtName:String?
    var TextNormalValue:String?
    var TextValue:String?
    var GCResultSummary:String?
    var ResultSummary:String?
    var Remarks:String?
    
    override func getPrimaryKey() -> [String]{
        return ["LaboratoryResultDtID"];
    }
}

public class LaboratoryResultDtDao{
    private var helper:DBHelper;
    private let p_LaboratoryResultDtID = "@p_LaboratoryResultDtID";
    
    public init() {
        helper = DBHelper();
    }
    public func get(LaboratoryResultDtID:Int) -> LaboratoryResultDt?{
        sharedInstance.database!.open()
        var query = helper.getRecord(tableName: "LaboratoryResultDt", lstPrimaryKey: LaboratoryResultDt().getPrimaryKey());
        query = query.replacingOccurrences(of: p_LaboratoryResultDtID, with: String(LaboratoryResultDtID));
        let row = DaoBase.getInstance().getDataRow(query: query);
        let temp = helper.dataListRowToObject(row: row, obj: LaboratoryResultDt());
        sharedInstance.database!.close()
        if(temp != nil) {
            let result = temp as! LaboratoryResultDt?;
            return result;
        }
        return nil
    }
    public func insert(record:LaboratoryResultDt) -> Bool{
        let query = helper.insert(tableName: "LaboratoryResultDt", record: record);
        return DaoBase.getInstance().executeNonQuery(query: query!);
    }
    public func update(record:LaboratoryResultDt) -> Bool{
        let query = helper.update(tableName: "LaboratoryResultDt", record: record);
        return DaoBase.getInstance().executeNonQuery(query: query!);
    }
    public func delete(LaboratoryResultDtID:Int) -> Bool{
        let record = get(LaboratoryResultDtID: LaboratoryResultDtID);
        let query = helper.delete(tableName: "LaboratoryResultDt", record: record!);
        return DaoBase.getInstance().executeNonQuery(query: query!);
    }
}

public class LaboratoryResultHd : BaseClass{
    var ID:NSNumber?
    var MRN:NSNumber?
    var ReferenceNo:String?
    var ResultDate:DateTime?
    var ResultTime:String?
    var ProviderName:String?
    var ParamedicName:String?
    var Remarks:String?
    var LastUpdatedDate:DateTime?
    
    override func getPrimaryKey() -> [String]{
        return ["ID"];
    }
}

public class LaboratoryResultHdDao{
    private var helper:DBHelper;
    private let p_ID = "@p_ID";
    
    public init() {
        helper = DBHelper();
    }
    public func get(ID:Int) -> LaboratoryResultHd?{
        sharedInstance.database!.open()
        var query = helper.getRecord(tableName: "LaboratoryResultHd", lstPrimaryKey: LaboratoryResultHd().getPrimaryKey());
        query = query.replacingOccurrences(of: p_ID, with: String(ID));
        let row = DaoBase.getInstance().getDataRow(query: query);
        let temp = helper.dataListRowToObject(row: row, obj: LaboratoryResultHd());
        sharedInstance.database!.close()
        if(temp != nil) {
            let result = temp as! LaboratoryResultHd?;
            return result;
        }
        return nil
    }
    public func insert(record:LaboratoryResultHd) -> Bool{
        let query = helper.insert(tableName: "LaboratoryResultHd", record: record);
        return DaoBase.getInstance().executeNonQuery(query: query!);
    }
    public func update(record:LaboratoryResultHd) -> Bool{
        let query = helper.update(tableName: "LaboratoryResultHd", record: record);
        return DaoBase.getInstance().executeNonQuery(query: query!);
    }
    public func delete(ID:Int) -> Bool{
        let record = get(ID: ID);
        let query = helper.delete(tableName: "LaboratoryResultHd", record: record!);
        return DaoBase.getInstance().executeNonQuery(query: query!);
    }
}

public class Patient : BaseClass{
    var MRN:NSNumber?;
    var MedicalNo:String?;
    var FullName:String?;
    var PreferredName:String?;
    var CityOfBirth:String?;
    var DateOfBirth:DateTime?;
    var GCSex:String?;
    var Sex:String?;
    var Gender:String?;
    var BloodType:String?;
    var BloodRhesus:String?;
    var EmailAddress:String?;
    var EmailAddress2:String?;
    var MobilePhoneNo1:String?;
    var MobilePhoneNo2:String?;
    var LastSyncDateTime:DateTime?;
    var LastSyncAppointmentDateTime:DateTime?;
    var LastSyncVaccinationDateTime:DateTime?;
    var LastSyncLabResultDateTime:DateTime?;
    public func getMobilePhoneNoDisplay() -> String{
        if (self.MobilePhoneNo2 != ""){
            return MobilePhoneNo1! + " / " + MobilePhoneNo2!;
        }
        return MobilePhoneNo1!;
    }
    public func getEmailAddressDisplay() -> String{
        if (self.EmailAddress2 != ""){
            return EmailAddress! + " / " + EmailAddress2!;
        }
        return EmailAddress!;
    }

    override func getPrimaryKey() -> [String]{
        return ["MRN"];
    }
}
public class PatientDao{
    private var helper:DBHelper;
    private let p_MRN = "@p_MRN";
    
    public init() {
        helper = DBHelper();
    }
    public func get(MRN:Int) -> Patient?{
        sharedInstance.database!.open()
        var query = helper.getRecord(tableName: "Patient", lstPrimaryKey: Patient().getPrimaryKey());
        query = query.replacingOccurrences(of: p_MRN, with: String(MRN));
        let row = DaoBase.getInstance().getDataRow(query: query);
        let temp = helper.dataListRowToObject(row: row, obj: Patient());
        sharedInstance.database!.close()
        if(temp != nil) {
            let result = temp as! Patient?;
            return result;
        }
        return nil
    }
    public func insert(record:Patient) -> Bool{
        let query = helper.insert(tableName: "Patient", record: record);
        return DaoBase.getInstance().executeNonQuery(query: query!);
    }
    public func update(record:Patient) -> Bool{
        let query = helper.update(tableName: "Patient", record: record);
        return DaoBase.getInstance().executeNonQuery(query: query!);
    }
    public func delete(MRN:Int) -> Bool{
        let record = get(MRN: MRN);
        let query = helper.delete(tableName: "Patient", record: record!);
        return DaoBase.getInstance().executeNonQuery(query: query!);
    }
}


public class Variable{
    var Code:String;
    var Value:String;
    
    public init (Code:String, Value:String){
        self.Code = Code;
        self.Value = Value;
    }
}

public class Setting : BaseClass{
    var SettingCode:String?
    var SettingName:String?
    var SettingValue:String?
    
    override func getPrimaryKey() -> [String]{
        return ["SettingCode"];
    }
    override func getNullableColumn() -> [String] {
        return ["SettingValue"];
    }


}

public class SettingDao{
    private var helper:DBHelper;
    private let p_SettingCode = "@p_SettingCode";
    
    public init() {
        helper = DBHelper();
    }
    public func get(SettingCode:String) -> Setting?{
        sharedInstance.database!.open()
        var query = helper.getRecord(tableName: "Setting", lstPrimaryKey: Setting().getPrimaryKey());
        query = query.replacingOccurrences(of: p_SettingCode, with: SettingCode);
        let row = DaoBase.getInstance().getDataRow(query: query);
        let temp = helper.dataListRowToObject(row: row, obj: Setting());
        sharedInstance.database!.close()
        if(temp != nil) {
            let result = temp as! Setting?;
            return result;
        }
        return nil    }
    public func insert(record:Setting) -> Bool{
        let query = helper.insert(tableName: "Setting", record: record);
        return DaoBase.getInstance().executeNonQuery(query: query!);
    }
    public func update(record:Setting) -> Bool{
        let query = helper.update(tableName: "Setting", record: record);
        return DaoBase.getInstance().executeNonQuery(query: query!);
    }
    public func delete(SettingCode:String) -> Bool{
        //let record:Setting = get(SettingCode);
        var query = helper.getRecord(tableName: "Setting", lstPrimaryKey: Setting().getPrimaryKey());
        query = query.replacingOccurrences(of: p_SettingCode, with: SettingCode);
        return DaoBase.getInstance().executeNonQuery(query: query);
    }
}

public class VaccinationShotDt : BaseClass{
    var `Type`:NSNumber?;
    var ID:NSNumber?;
    var VaccinationDate:DateTime?;
    var MRN:NSNumber?;
    var ParamedicName:String?;
    var VaccinationTypeID:NSNumber?;
    var VaccinationTypeName:String?;
    var VaccinationNo:String?;
    var VaccineName:String?;
    var Dose:NSNumber?;
    var DoseUnit:String?;
    var LastUpdatedDate:DateTime?
    
    override func getPrimaryKey() -> [String]{
        return ["Type", "ID"];
    }
}
public class VaccinationShotDtDao{
    private var helper:DBHelper;
    private let p_Type = "@p_Type";
    private let p_ID = "@p_ID";
    
    public init() {
        helper = DBHelper();
    }
    public func get(Type:Int, ID:Int) -> VaccinationShotDt?{
        sharedInstance.database!.open()
        var query = helper.getRecord(tableName: "VaccinationShotDt", lstPrimaryKey: VaccinationShotDt().getPrimaryKey());
        query = query.replacingOccurrences(of: p_Type, with: String(Type));
        query = query.replacingOccurrences(of: p_ID, with: String(ID));
        let row = DaoBase.getInstance().getDataRow(query: query);
        let temp = helper.dataListRowToObject(row: row, obj: VaccinationShotDt());
        sharedInstance.database!.close()
        if(temp != nil) {
            let result = temp as! VaccinationShotDt?;
            return result;
        }
        return nil
    }
    public func insert(record:VaccinationShotDt) -> Bool{
        let query = helper.insert(tableName: "VaccinationShotDt", record: record);
        return DaoBase.getInstance().executeNonQuery(query: query!);
    }
    public func update(record:VaccinationShotDt) -> Bool{
        let query = helper.update(tableName: "VaccinationShotDt", record: record);
        return DaoBase.getInstance().executeNonQuery(query: query!);
    }
    public func delete(Type:Int, ID:Int) -> Bool{
        let record = get(Type: Type, ID: ID);
        let query = helper.delete(tableName: "VaccinationShotDt", record: record!);
        return DaoBase.getInstance().executeNonQuery(query: query!);
    }
}


public class vAppointment : BaseClass{
    var AppointmentID:NSNumber?
    var MRN:NSNumber?
    var FullName:String?
    var ServiceUnitName:String?
    var QueueNo:NSNumber?
    var StartDate:DateTime?
    var ReminderDate:DateTime?
    var EndDate:DateTime?
    var StartTime:String?
    var EndTime:String?
    var cfStartTime:String?
    var VisitTypeName:String?
    var ParamedicName:String?
    var SpecialtyName:String?
    var GCAppointmentStatus:String?
    var LastUpdatedDate:DateTime?
}

public class vVaccinationType : BaseClass{
    var VaccinationTypeID:NSNumber?
    var VaccinationTypeName:String?
}

File: OttimoSupport/app/src/main/java/samanasoft/android/ottimo/dal/DataLayer.java
package samanasoft.android.ottimo.dal;

import android.content.Context;
import android.database.Cursor;
import samanasoft.android.framework.DateTime;
import samanasoft.android.framework.DbHelper;
import samanasoft.android.framework.DaoBase;
import samanasoft.android.framework.Helper;
import samanasoft.android.framework.attribute.Column;
import samanasoft.android.framework.attribute.Table;
import samanasoft.android.framework.attribute.EnumAttribute.Bool;
import samanasoft.android.framework.attribute.EnumAttribute.DataType;
import samanasoft.android.ottimo.common.Constant;

public class DataLayer{
    //region Variable
    public static class Variable {
        public int Code;
        public String Value;

        public Variable(int Code, String Value) {
            this.Code = Code;
            this.Value = Value;
        }
    }
    //endregion
    //region Appointment
    @Table(Name = "Appointment")
    public static class Appointment{
        @Column(DataType = DataType.INT, Name = "AppointmentID", IsPrimaryKey = Bool.TRUE)
        public int AppointmentID;

        @Column(DataType = DataType.STRING, Name = "PatientName")
        public String PatientName;

        @Column(DataType = DataType.STRING, Name = "PreferredName")
        public String PreferredName;

        @Column(DataType = DataType.STRING, Name = "ServiceUnitName")
        public String ServiceUnitName;

        @Column(DataType = DataType.INT, Name = "QueueNo")
        public int QueueNo;

        @Column(DataType = DataType.DATETIME, Name = "StartDate")
        public DateTime StartDate;

        @Column(DataType = DataType.DATETIME, Name = "EndDate")
        public DateTime EndDate;

        @Column(DataType = DataType.STRING, Name = "StartTime")
        public String StartTime;

        @Column(DataType = DataType.STRING, Name = "EndTime")
        public String EndTime;

        @Column(DataType = DataType.STRING, Name = "cfStartTime")
        public String cfStartTime;

        @Column(DataType = DataType.STRING, Name = "VisitTypeName")
        public String VisitTypeName;

        @Column(DataType = DataType.STRING, Name = "ParamedicName")
        public String ParamedicName;

        @Column(DataType = DataType.STRING, Name = "SpecialtyName")
        public String SpecialtyName;

        @Column(DataType = DataType.STRING, Name = "MobilePhoneNo")
        public String MobilePhoneNo;

        @Column(DataType = DataType.STRING, Name = "MobilePhoneNo2")
        public String MobilePhoneNo2;

        @Column(DataType = DataType.STRING, Name = "EmailAddress")
        public String EmailAddress;

        @Column(DataType = DataType.STRING, Name = "SMSLogStatus")
        public String SMSLogStatus;

        @Column(DataType = DataType.INT, Name = "NoOfMessageSent")
        public int NoOfMessageSent;

        @Column(DataType = DataType.DATETIME, Name = "LastUpdatedDate")
        public DateTime LastUpdatedDate;

        public Boolean IsChecked = false;
        public String getMobilePhoneNoDisplay(){
            if(MobilePhoneNo2 != null && !MobilePhoneNo2.isEmpty())
                return MobilePhoneNo + "/" + MobilePhoneNo2;
            return MobilePhoneNo;
        }
    }
    public static class AppointmentDao{
        private DbHelper helper;
        private DaoBase daoBase;
        private final String p_AppointmentID = "@p_AppointmentID";

        public AppointmentDao(Context context){
            this.helper = new DbHelper(Appointment.class);
            this.daoBase = new DaoBase(context);
        }
        public Appointment get(int AppointmentID){
            String query = helper.getRecord();
            query = query.replace(p_AppointmentID, Integer.toString(AppointmentID));
            Cursor row = daoBase.getDataRow(query);
            return (row == null) ? null : (Appointment)helper.dataRowToObject(row, new Appointment());
        }
        public int insert(Appointment record){
            String query = helper.insert(record);
            return daoBase.executeNonQuery(query);
        }
        public int update(Appointment record){
            String query = helper.update(record);
            return daoBase.executeNonQuery(query);
        }
        public int delete(int AppointmentID){
            Appointment record = get(AppointmentID);
            String query = helper.delete(record);
            return daoBase.executeNonQuery(query);
        }
    }
    //endregion
    //region MessageLog
    @Table(Name = "MessageLog")
    public static class MessageLog{
        @Column(DataType = DataType.INT, Name = "ID")
        public int ID;

        @Column(DataType = DataType.DATETIME, Name = "LogDate")
        public DateTime LogDate;

        @Column(DataType = DataType.STRING, Name = "LogTime")
        public String LogTime;

        @Column(DataType = DataType.STRING, Name = "PatientName")
        public String PatientName;

        @Column(DataType = DataType.STRING, Name = "MobilePhoneNo")
        public String MobilePhoneNo;

        @Column(DataType = DataType.STRING, Name = "MessageText")
        public String MessageText;
    }

    public static class MessageLogDao{
        private DbHelper helper;
        private DaoBase daoBase;
        private final String p_ID = "@p_ID";

        public MessageLogDao(Context context){
            this.helper = new DbHelper(MessageLog.class);
            this.daoBase = new DaoBase(context);
        }
        public MessageLog get(int ID){
            String query = helper.getRecord();
            query = query.replace(p_ID, Integer.toString(ID));
            Cursor row = daoBase.getDataRow(query);
            return (row == null) ? null : (MessageLog)helper.dataRowToObject(row, new MessageLog());
        }
        public int insert(MessageLog record){
            String query = helper.insert(record);
            return daoBase.executeNonQuery(query);
        }
        public int update(MessageLog record){
            String query = helper.update(record);
            return daoBase.executeNonQuery(query);
        }
        public int delete(int ID){
            MessageLog record = get(ID);
            String query = helper.delete(record);
            return daoBase.executeNonQuery(query);
        }
    }
    //endregion

    //region Paramedic
    @Table(Name = "Paramedic")
    public static class Paramedic{
        @Column(DataType = DataType.INT, Name = "ParamedicID")
        public Integer ParamedicID;
        @Column(DataType = DataType.STRING, Name = "ParamedicName")
        public String ParamedicName;
    }
    //endregion

	//region Menu
	@Table(Name = "Setting")
	public static class Setting{
		@Column(DataType = DataType.STRING, Name = "SettingCode", IsPrimaryKey = Bool.TRUE)
		public String SettingCode;

		@Column(DataType = DataType.STRING, Name = "SettingName")
		public String SettingName;

		@Column(DataType = DataType.STRING, Name = "SettingValue")
		public String SettingValue;
	}

	public static class SettingDao{
		private DbHelper helper;
		private DaoBase daoBase;
        private final String p_SettingCode = "@p_SettingCode";

		public SettingDao(Context context){
			this.helper = new DbHelper(Setting.class);
			this.daoBase = new DaoBase(context);
		}
		public Setting get(String SettingCode){
			String query = helper.getRecord();
			query = query.replace(p_SettingCode, SettingCode);
			Cursor row = daoBase.getDataRow(query);
			return (row == null) ? null : (Setting)helper.dataRowToObject(row, new Setting());
		}
		public int insert(Setting record){
			String query = helper.insert(record);
			return daoBase.executeNonQuery(query);
		}
		public int update(Setting record){
			String query = helper.update(record);
			return daoBase.executeNonQuery(query);
		}
		public int delete(String SettingCode){
            Setting record = get(SettingCode);
			String query = helper.delete(record);
			return daoBase.executeNonQuery(query);
		}
	}
	//endregion

    //region TemplateText
    @Table(Name = "TemplateText")
    public static class TemplateText{
        @Column(DataType = DataType.INT, Name = "TemplateID")
        public Integer TemplateID;
        @Column(DataType = DataType.STRING, Name = "TemplateName")
        public String TemplateName;
        @Column(DataType = DataType.STRING, Name = "TemplateContent")
        public String TemplateContent;
    }
    //endregion

    //region VaccinationType
    @Table(Name = "VaccinationType")
    public static class VaccinationType{
        @Column(DataType = DataType.INT, Name = "VaccinationTypeID")
        public Integer VaccinationTypeID;
        @Column(DataType = DataType.STRING, Name = "VaccinationTypeCode")
        public String VaccinationTypeCode;
        @Column(DataType = DataType.STRING, Name = "VaccinationTypeName")
        public String VaccinationTypeName;
    }
    //endregion
}
